#161342846 Users should be able to receive notifications

ucheg6 opened the pull request:
#### What does this PR do?
-  Users should be able to receive notifications

#### Description of Task to be completed?
- Update Users model and create a new migration file to add allowNotifications 
- Create user endpoint to allow users to opt in/ out of notifications
- Create a notification model and migration file
- Create service workers to intercept new articles, reactions, comments and update the notifications table
- Create a worker to dispatch email notifications
- Wrote tests for all functionalities

#### How should this be manually tested?
- Git clone the repository
- cd into the `haven-ah-backend` folder
- run `npm run start-dev`
- Fire up the route `localhost:5000/api/v1/articles` as a post request
- To test the entire application, run `npm run test`

#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
[#161342846](https://www.pivotaltracker.com/story/show/161342846)

#### Screenshots (if appropriate)
![authorhavennotification](https://user-images.githubusercontent.com/41582565/48150345-ac77dc00-e2be-11e8-93a3-3a81c748d58b.png)

<img width="834" alt="screen shot 2018-11-08 at 6 46 20 pm" src="https://user-images.githubusercontent.com/34472497/48221239-d221e580-e391-11e8-8983-523e229a3b64.png">
<img width="1112" alt="screen shot 2018-11-08 at 8 03 23 pm" src="https://user-images.githubusercontent.com/34472497/48221356-2200ac80-e392-11e8-8d1d-417f8b86edef.png">
<img width="719" alt="screen shot 2018-11-09 at 7 06 47 am" src="https://user-images.githubusercontent.com/34472497/48247124-720f5b80-e3f2-11e8-939e-0d6d157a0543.png">
<img width="598" alt="screen shot 2018-11-09 at 6 07 09 am" src="https://user-images.githubusercontent.com/34472497/48247271-f7930b80-e3f2-11e8-96f2-3e819ed213d2.png">

TheoOkafor reviewed (commented):
One word: Impressive!

PriscillaSam commented on server/repository/reactionRepository.js:
```diff
@@ -0,0 +1,91 @@
+      articleId,
+      userId
+    });
+    notificationRepo.createNotification({
```
@ProfJigsaw Usually the idea of a repository is to have proper separation of concerns between the models. I don't think we should call the notification repo here. We can call this in the reaction controller and pass in the values to create the notification.

victor-nweze-deel replied:
@PriscillaSam Yes, I thought of that but if I remove this one-time call from the repo then I have to sacrifice DRY in the controllers. This is because there are several cases on creating reactions and comments and each one would need a call to createNotification. This would be more bogus.

PriscillaSam replied:
@ProfJigsaw I don't think so because this particular call here is just to notify when a reaction is made. Making this call on the controller doesn't sacrifice DRY in any way.

victor-nweze-deel replied:
Fixed. Thanks for the catch

PriscillaSam commented on server/models/notifications.js:
```diff
@@ -0,0 +1,34 @@
+  }, {});
+  Notifications.associate = (models) => {
+    Notifications.belongsTo(models.Articles, {
+      foreignKey: 'articleid',
```
Hey, guys shouldn't this be `articleId`?

victor-nweze-deel replied:
Noted!

PriscillaSam commented on server/services/notifiers/newArticleNotifier.js:
```diff
@@ -0,0 +1,41 @@
+  );
+  const mailInformation = await emailer.sendEmail(mailOptions);
+  return mailInformation;
+};
```
Very great work done here. Well done guys.

PriscillaSam commented on server/controllers/comment.js:
```diff
@@ -48,6 +49,12 @@ class Comment {
+    });
     return goodHttpResponse(response, 201, 'Reply created', newReply);
   }
 
```
Great @ProfJigsaw. Looks good here. Well done.

seyi-adeleke commented on server/repository/notificationRepository.js:
```diff
@@ -0,0 +1,40 @@
+    let content;
+    let newNotification = {};
+    switch (notification.type) {
+      case 'NEW_ARTICLE_UPDATE':
```
These should be constants exported into the file `NEW_ARTICLE_UPDATE `

victor-nweze-deel replied:
Noted and fixed.

victor-nweze-deel committed:
feat(notifications): Users should be able to receive notifications

- Update user model and create migration file to add column allowNotifications
- Create notification model and migration file
- Create serviceWorkers to intercept new articles and update the notifications table
- Create worker to dispatch email notifications
- Update swagger.js file
